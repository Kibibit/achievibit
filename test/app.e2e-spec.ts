import { NestExpressApplication } from '@nestjs/platform-express';
import { Test, TestingModule } from '@nestjs/testing';
import nunjucks from 'nunjucks';
import request from 'supertest';

import { AppModule } from '@kb-app';
import { InMemoryDatabaseModule, pullRequestCreatedEvent, webhookPingEvent } from '@kb-dev-tools';

// new ConfigService({});

describe('AppController (e2e)', () => {
  let app: NestExpressApplication;

  beforeEach(async () => {
    const moduleFixture: TestingModule = await Test.createTestingModule({
      imports: [ AppModule ]
    }).compile();

    app = moduleFixture.createNestApplication();
    app.setViewEngine('njk');
    app.set('view engine', 'njk');
    nunjucks.configure('views', {
      autoescape: true,
      express: app,
      watch: true
    });

    await app.init();
  });

  test('/ (GET) should return homepage at root of App', async () => {
    console.log('server is running');
    const response = await request(app.getHttpServer())
      .get('/')
      .expect(200);

    expect(response.text).toMatchSnapshot();
  });

  test('/api (GET) should return API hello page with package info', async () => {
    console.log('server is running');
    const response = await request(app.getHttpServer())
      .get('/api')
      .expect(200);

    expect(response.text).toMatchSnapshot();
  });

  afterEach(async () => {
    await InMemoryDatabaseModule.closeDatabase();
  });

  test('/ (POST) from github ping event should create db repo', async () => {
    const server = app.getHttpServer();
    const sendWebhookResponse = await request(server)
      .post('/')
      .send(webhookPingEvent.payload) // x-www-form-urlencoded upload
      .set('Accept', 'application/json')
      .set('x-github-event', webhookPingEvent.event)
      .expect(201);

    expect(sendWebhookResponse.text).toMatchSnapshot();

    const getAllReposResponse = await request(server)
      .get('/api/repos')
      .expect(200);

    expect(getAllReposResponse.body).toMatchSnapshot();
  });

  test('/ (POST) from github pull request created event should create user', async () => {
    const server = app.getHttpServer();
    const sendWebhookResponse = await request(server)
      .post('/')
      .set('Accept', 'application/json')
      .set('x-github-event', pullRequestCreatedEvent.event)
      .send(pullRequestCreatedEvent.payload)
      .expect(201);

    expect(sendWebhookResponse.text).toMatchSnapshot();

    const getAllUsersResponse = await request(server)
      .get('/api/users')
      .expect(200);

    expect(getAllUsersResponse.body).toMatchSnapshot();
  });
});
