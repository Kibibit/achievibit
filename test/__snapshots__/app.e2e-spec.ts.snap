// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AppController (e2e) / (GET) should return homepage at root of App 1`] = `
"<!DOCTYPE html>
<html lang=\\"en\\">

<head>
  <meta http-equiv=\\"Content-Type\\" content=\\"text/html; charset=UTF-8\\" />
  <meta name=\\"viewport\\" content=\\"width=device-width, initial-scale=1, maximum-scale=1.0, user-scalable=no\\" />
  <meta name=\\"description\\" content=\\"Achievements for your GitHub Repositories\\">
  <meta name=\\"viewport\\" content=\\"width=device-width, initial-scale=1.0, minimum-scale=1.0\\">
  <meta property=\\"og:title\\" content=\\"achievibit - by kibibit\\">
  <meta property=\\"og:image\\" content=\\"https://achievibit.kibibit.io/images/favicon.png\\">
  <meta property=\\"og:description\\" content=\\"Achievements for your GitHub Repositories\\">
  <meta property=\\"og:url\\" content=\\"https://achievibit.kibibit.io/\\" />
  <meta property=\\"og:site_name\\" content=\\"achievibit\\" />
  <title>achievibit - from kibibit</title>

  <!-- CSS  -->
  <link href=\\"https://fonts.googleapis.com/icon?family=Material+Icons\\" rel=\\"stylesheet\\">
  <link href=\\"public/materialize.css\\" type=\\"text/css\\" rel=\\"stylesheet\\" media=\\"screen,projection\\" />
  <link href=\\"public/homepage.css\\" type=\\"text/css\\" rel=\\"stylesheet\\" media=\\"screen,projection\\" />

  <!-- Global Site Tag (gtag.js) - Google Analytics -->
  <script async src=\\"https://www.googletagmanager.com/gtag/js?id=UA-108880116-1\\"></script>
  <script>
    window.dataLayer = window.dataLayer || [];

    function gtag() {
      dataLayer.push(arguments);
    }
    gtag('js', new Date());

    gtag('config', 'UA-108880116-1', {
            'transport_type': 'beacon',
            'custom_map': {
                'dimension1': 'url',
                'dimension2': 'name'
            }
        });
  </script>
  <script>
    gtag('create', 'UA-108880116-1', 'auto');
    /**
     * Function that tracks a click on an outbound link in Analytics.
     * This function takes a valid URL string as an argument, and uses that URL string
     * as the event label. Setting the transport method to 'beacon' lets the hit be sent
     * using 'navigator.sendBeacon' in browser that support it.
     */
    var trackOutboundLink = function (url, name) {
      let params = {
        'event_category': 'click',
        'url': url,
        'event_callback': function () {
          document.location = url;
        }
      };

      if (name) {
        params.name = name;
      }

      gtag('event', 'outbound', params);
    }
  </script>
</head>

<body>
  <nav class=\\"white\\" role=\\"navigation\\">
    <div class=\\"nav-wrapper container\\">
      <a id=\\"logo-container\\" href=\\"#\\" class=\\"brand-logo\\">
        <img src=\\"public/images/favicon.png\\" width=\\"50\\" height=\\"50\\">
      </a>
      <ul class=\\"right hide-on-med-and-down\\">
        <li>
          <a href=\\"#\\">
            <i class=\\"material-icons\\">search</i>
          </a>
        </li>
      </ul>

      <ul id=\\"nav-mobile\\" class=\\"side-nav\\">
        <li>
          <a href=\\"#\\">Search&nbsp;
            <i class=\\"material-icons\\">search</i>
          </a>
        </li>
      </ul>
      <a href=\\"#\\" data-activates=\\"nav-mobile\\" class=\\"button-collapse\\">
        <i class=\\"material-icons\\">menu</i>
      </a>
    </div>
  </nav>

  <div id=\\"index-banner\\" class=\\"parallax-container\\">
    <div class=\\"section no-pad-bot\\">
      <div class=\\"container\\">
        <br>
        <br>
        <h1 class=\\"header center main-color text-lighten-2\\" style=\\"display: flex; justify-content: center; align-items: center\\">
          <img src=\\"public/images/unlocked.png\\" width=\\"150\\" height=\\"150\\">
          <span>achievibit</span>
        </h1>
        <div class=\\"row center\\">
          <h5 class=\\"header col s12 light\\">
            Achievements system as a GitHub WebHook
          </h5>
        </div>
        <div class=\\"row center\\">
          <a href=\\"https://www.github.com/Kibibit/achievibit\\" onclick=\\"trackOutboundLink('https://www.github.com/Kibibit/achievibit'); return false;\\"
            id=\\"download-button\\" class=\\"btn-large waves-effect waves-light teal lighten-1\\">Integrate in your repository</a>
        </div>
        <br>
        <br>

      </div>
    </div>
    <div class=\\"parallax\\">
      <img src=\\"public/images/background1.png\\" alt=\\"Unsplashed background img 1\\">
    </div>
  </div>


  <div class=\\"container\\">
    <div class=\\"section\\">

      <!--   Icon Section   -->
      <div class=\\"row\\">
        <div class=\\"col s12 m4\\">
          <div class=\\"icon-block\\" style=\\"display: flex; flex-direction: column;\\">
            <h2 class=\\"center brown-text\\">
              <i class=\\"material-icons\\">verified_user</i>
            </h2>
            <h5 class=\\"center\\">Getting achievements</h5>

            <p class=\\"light center\\">
              achievibit adds achievements to each pull request you merge. In order to get achievements, just merge your pull requests
              like you normally do. achievibit will grant achievements based on some data achievibit listens to.
            </p>
          </div>
        </div>

        <div class=\\"col s12 m4\\">
          <div class=\\"icon-block\\" style=\\"display: flex; flex-direction: column;\\">
            <h2 class=\\"center brown-text\\">
              <i class=\\"material-icons\\">flash_on</i>
            </h2>
            <h5 class=\\"center\\">See achievements on GitHub</h5>

            <p class=\\"light center\\">
              We also have a Chrome that will integrate achievements right into GitHub's profiles
            </p>
            <a target=\\"_blank\\" href=\\"https://chrome.google.com/webstore/detail/achievibit/iddkmddomdohnihbehiamfnmpomlhpee\\" onclick=\\"trackOutboundLink('https://chrome.google.com/webstore/detail/achievibit/iddkmddomdohnihbehiamfnmpomlhpee'); return false;\\"
              class=\\"btn-large waves-effect waves-light teal lighten-1 center\\">Download Extension</a>
          </div>
        </div>

        <div class=\\"col s12 m4\\">
          <div class=\\"icon-block\\" style=\\"display: flex; flex-direction: column;\\">
            <h2 class=\\"center brown-text\\">
              <i class=\\"material-icons\\">create</i>
            </h2>
            <h5 class=\\"center\\">Create unique achievements</h5>

            <p class=\\"light center\\">
              you can create your own achievements for
              <strong>achievibit</strong>.
            </p>
          </div>
        </div>
      </div>

    </div>
  </div>


  <div class=\\"parallax-container valign-wrapper\\">
    <div class=\\"section no-pad-bot\\">
      <div class=\\"container\\">
        <div class=\\"row center\\">
          <h5 class=\\"header col s12 light\\">Contribute to get achievements!</h5>
        </div>
      </div>
    </div>
    <div class=\\"parallax\\">
      <img src=\\"public/images/background2.png\\" alt=\\"Unsplashed background img 2\\">
    </div>
  </div>

  <div class=\\"container\\">
    <div class=\\"section\\">

      <div class=\\"row\\">
        <div class=\\"col s12 center\\">
          <h3>
            <i class=\\"mdi-content-send brown-text\\"></i>
          </h3>
          <h4>These repositories have achievibit integrated</h4>
          <div class=\\"list-container\\">
            
          </div>
        </div>
      </div>

    </div>
  </div>

  <div class=\\"container\\">
    <div class=\\"section\\">

      <div class=\\"row\\">
        <div class=\\"col s12 center\\">
          <h3>
            <i class=\\"mdi-content-send brown-text\\"></i>
          </h3>
          <h4> users working on achievements!</h4>
          <div class=\\"list-container\\">
            
          </div>
        </div>
      </div>

    </div>
  </div>

  <div class=\\"container\\">
    <div class=\\"section\\">

      <div class=\\"row\\">
        <div class=\\"col s12 center\\">
          <h3>
            <i class=\\"mdi-content-send brown-text\\"></i>
          </h3>
          <h4> organizations that use achievibit</h4>
          <div class=\\"list-container\\">
            
          </div>
        </div>
      </div>

    </div>
  </div>


  <div class=\\"parallax-container valign-wrapper\\">
    <div class=\\"section no-pad-bot\\">
      <div class=\\"container\\">
        <div class=\\"row center\\">
          <h5 class=\\"header col s12 light\\">Join now and start achieving</h5>
        </div>
      </div>
    </div>
    <div class=\\"parallax\\">
      <img src=\\"public/images/background3.png\\" alt=\\"Unsplashed background img 3\\">
    </div>
  </div>

  <footer class=\\"page-footer teal\\">
    <div class=\\"container\\">
      <div class=\\"row\\">
        <div class=\\"col l6 s12\\">
          <h5 class=\\"white-text\\">kibibit</h5>
          <p class=\\"grey-text text-lighten-4\\">
            kibibit is an open source team of software developers, trying to make the world easier piece by piece.

            <strong>you're more than welcome to contribute!</strong>
          </p>


        </div>
        <div class=\\"col l3 s12\\">
        </div>
        <div class=\\"col l3 s12\\">
          <h5 class=\\"white-text\\">Sharing the love</h5>
          <ul>
            <li>
              <a class=\\"white-text\\" href=\\"https://heroku.com/\\" onclick=\\"trackOutboundLink('https://heroku.com/'); return false;\\">Heroku</a>
            </li>
            <li>
              <a class=\\"white-text\\" href=\\"https://mlab.com/\\" onclick=\\"trackOutboundLink('https://mlab.com/'); return false;\\">mLab</a>
            </li>
            <li>
              <a class=\\"white-text\\" href=\\"https://getmdl.io/\\" onclick=\\"trackOutboundLink('https://getmdl.io/'); return false;\\">Material lite</a>
            </li>
            <li>
              <a class=\\"white-text\\" href=\\"http://materializecss.com\\" onclick=\\"trackOutboundLink('http://materializecss.com'); return false;\\">Materialize</a>
            </li>
          </ul>
        </div>
      </div>
    </div>
    <div class=\\"footer-copyright\\">
      <div class=\\"container\\">
        <!-- Made by <a class=\\"brown-text text-lighten-3\\" href=\\"http://materializecss.com\\">Materialize</a> -->
      </div>
    </div>
  </footer>


  <!--  Scripts-->
  <script src=\\"https://code.jquery.com/jquery-2.1.1.min.js\\"></script>
  <script src=\\"https://cdnjs.cloudflare.com/ajax/libs/trianglify/1.0.1/trianglify.min.js\\"></script>
  <script src=\\"public/materialize.js\\"></script>
  <script src=\\"public/homepage-init.js\\"></script>

</body>

</html>
"
`;

exports[`AppController (e2e) / (POST) from github ping event should create db repo 1`] = `"NewConnection"`;

exports[`AppController (e2e) / (POST) from github ping event should create db repo 2`] = `
Array [
  Object {
    "fullname": "Thatkookooguy/test-new-achievibit-events",
    "name": "test-new-achievibit-events",
    "url": "https://github.com/Thatkookooguy/test-new-achievibit-events",
  },
]
`;

exports[`AppController (e2e) / (POST) from github pull request created event should create user 1`] = `"PullRequestOpened"`;

exports[`AppController (e2e) / (POST) from github pull request created event should create user 2`] = `Array []`;

exports[`AppController (e2e) / (POST) from github pull request should create db items 1`] = `"NewConnection"`;

exports[`AppController (e2e) / (POST) from github pull request should create db items 2`] = `
Array [
  Object {
    "fullname": "Thatkookooguy/test-new-achievibit-events",
    "name": "test-new-achievibit-events",
    "url": "https://github.com/Thatkookooguy/test-new-achievibit-events",
  },
]
`;

exports[`AppController (e2e) /api (GET) should return API hello page with package info 1`] = `"{\\"name\\":\\"achievibit\\",\\"version\\":\\"0.0.1\\",\\"description\\":\\"\\",\\"author\\":\\"\\",\\"license\\":\\"MIT\\"}"`;
